lightning/src/util/persist.rs

```diff
@@ -249,10 +258,821 @@ where
+			logger: &TestLogger::new(),
+			maximum_pending_updates: 11,
+			entropy_source: &chanmon_cfgs[1].keys_manager,
+			signer_provider: &chanmon_cfgs[1].keys_manager,
+		};
+		let mut node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
+		let chain_mon_0 = test_utils::TestChainMonitor::new(
+			Some(&chanmon_cfgs[0].chain_source),
+			&chanmon_cfgs[0].tx_broadcaster,
+			&chanmon_cfgs[0].logger,
+			&chanmon_cfgs[0].fee_estimator,
+			&persister_0,
+			&chanmon_cfgs[0].keys_manager,
+		);
+		let chain_mon_1 = test_utils::TestChainMonitor::new(
+			Some(&chanmon_cfgs[1].chain_source),
+			&chanmon_cfgs[1].tx_broadcaster,
+			&chanmon_cfgs[1].logger,
+			&chanmon_cfgs[1].fee_estimator,
+			&persister_1,
+			&chanmon_cfgs[1].keys_manager,
+		);
+		node_cfgs[0].chain_monitor = chain_mon_0;
+		node_cfgs[1].chain_monitor = chain_mon_1;
+		let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
+		let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+
+		let broadcaster_0 = &chanmon_cfgs[2].tx_broadcaster;
+		let broadcaster_1 = &chanmon_cfgs[3].tx_broadcaster;
+
+		// Check that the persisted channel data is empty before any channels are
+		// open.
+		let mut persisted_chan_data_0 = persister_0
+			.read_channel_monitors_with_updates(&broadcaster_0, &chanmon_cfgs[0].fee_estimator)
+			.unwrap();
+		assert_eq!(persisted_chan_data_0.len(), 0);
+		let mut persisted_chan_data_1 = persister_1
+			.read_channel_monitors_with_updates(&broadcaster_1, &chanmon_cfgs[1].fee_estimator)
+			.unwrap();
+		assert_eq!(persisted_chan_data_1.len(), 0);
+
+		// Helper to make sure the channel is on the expected update ID.
+		macro_rules! check_persisted_data {
+			($expected_update_id: expr) => {
+				persisted_chan_data_0 = persister_0
+					.read_channel_monitors_with_updates(
+						&broadcaster_0,
+						&chanmon_cfgs[0].fee_estimator,
+					)
+					.unwrap();
+				assert_eq!(persisted_chan_data_0.len(), 1);
+				for (_, mon) in persisted_chan_data_0.iter() {
+					assert_eq!(mon.get_latest_update_id(), $expected_update_id);
+				}
+				persisted_chan_data_1 = persister_1
+					.read_channel_monitors_with_updates(
+						&broadcaster_1,
+						&chanmon_cfgs[1].fee_estimator,
+					)
+					.unwrap();
+				assert_eq!(persisted_chan_data_1.len(), 1);
+				for (_, mon) in persisted_chan_data_1.iter() {
+					assert_eq!(mon.get_latest_update_id(), $expected_update_id);
+				}
+			};
+		}
+
+		// Create some initial channel and check that a channel was persisted.
+		let _ = create_announced_chan_between_nodes(&nodes, 0, 1);
+		check_persisted_data!(0);
+
+		// Send a few payments and make sure the monitors are updated to the latest.
+		send_payment(&nodes[0], &vec![&nodes[1]][..], 8_000_000);
+		check_persisted_data!(EXPECTED_UPDATES_PER_PAYMENT);
+		send_payment(&nodes[1], &vec![&nodes[0]][..], 4_000_000);
+		check_persisted_data!(2 * EXPECTED_UPDATES_PER_PAYMENT);
+
+		// Send a few more payments.
+		let mut sender = 0;
+		for i in 3..=100 {
```

Comment:
Could we reduce the number of iterations? 100 takes over 4 seconds for the test to run on my M2 Pro.

Reply:
Huh. 1.5s on my M1 MBP. Am I running the tests without some kind of debugging flag? Anyway, I can reduce the number of iterations. My goal here was to size the pending update window as a prime larger than what we'd write for payments, and then iterate enough to have all the "alignments" with different phases of payments for both parties. I think that boils down to 14, which runs in ~.25s on my machine, so <1s on yours probably.

Reply:
Yeah, it's about 0.5 seconds using 14. I've been running:

```
cargo +1.54.0 test -p lightning -- persister_with_real_monitors
```

And down to 0.16 seconds when using rustc 1.70, or about a second with 100 iterations. So the version of rustc is partly to blame.

Reply:
Ah yeah, I abandoned trying to run 1.54 locally, I get some kind of error from a dependency... I probably just need to pin it, but that kind of stuff adds up so I just run 1.71. It runs in .16s, you say!? Well, now I want an M2.